### PR TITLE
Feature/tm lowess cv span

### PR DESCRIPTION
**Pull request by singjc**
## Description

This pull request introduces automatic LOWESS span selection via cross-validation to the `TransformationModelLowess` class, along with supporting utilities and statistical functions. The main goal was to add an option to automatically tune the smoothing span parameter using cross-validation for the lowess model. The changes also add a reusable cross-validation utility and extend statistical functions for error calculation and quantile computation. This is used for automating some other params in another separate incoming PR for OpenSwathWorkflow.

**Automatic LOWESS span selection and cross-validation:**

* Added logic to `TransformationModelLowess` to optionally select the LOWESS span parameter automatically using cross-validation, including new parameters for controlling grid search, metrics, and fold selection. [[1]](diffhunk://#diff-f9e0ac1599af18287764e21891ee77f85fc80fc1ed68234ab1a1b18e103e1847R80-R154) [[2]](diffhunk://#diff-f9e0ac1599af18287764e21891ee77f85fc80fc1ed68234ab1a1b18e103e1847R188-R216)
* Introduced helper functions to parse candidate span grids, build span grids with clamping and deduplication, and score residuals using different error metrics (`rmse`, `mae`, quantiles). These helpers are hidden from the public API. [[1]](diffhunk://#diff-4fe7180f7ee80dfc9279bae8aa9c7d248c3852d4cb089a0ced7601720a670f4eR48-R110) [[2]](diffhunk://#diff-f9e0ac1599af18287764e21891ee77f85fc80fc1ed68234ab1a1b18e103e1847R229-R305)

**Cross-validation and grid search utilities:**

* Added a new header-only utility `CrossValidation.h` providing deterministic K-fold/LOO split and a generic 1-D grid search function with flexible tie-breaking and scoring.
* Integrated the new cross-validation utility into the build system. [[1]](diffhunk://#diff-06f0dd8b504ee3a7d1f35d8ccdb950fa027b1da0f527b0c1d91daf8e57aaa7d2R1-R18) [[2]](diffhunk://#diff-b83abd395618ba2442c21bf6733afe6fe131eda2be60b0161e64ef7626bdb7acR15) [[3]](diffhunk://#diff-b83abd395618ba2442c21bf6733afe6fe131eda2be60b0161e64ef7626bdb7acR113)

**Statistical function enhancements:**

* Added `meanAbsoluteError`, `quantile` (with linear interpolation), and `rootMeanSquareError` functions to `OpenMS::Math::StatisticFunctions` to support new error metrics for model selection. [[1]](diffhunk://#diff-74d1036a733aa7c49ebc45f8478be9e763073146a3b07945536dcd17c4e389ccR200-R227) [[2]](diffhunk://#diff-74d1036a733aa7c49ebc45f8478be9e763073146a3b07945536dcd17c4e389ccR287-R352) [[3]](diffhunk://#diff-74d1036a733aa7c49ebc45f8478be9e763073146a3b07945536dcd17c4e389ccR488-R506)
* Included necessary headers and dependencies for new statistical and string operations. [[1]](diffhunk://#diff-74d1036a733aa7c49ebc45f8478be9e763073146a3b07945536dcd17c4e389ccR13) [[2]](diffhunk://#diff-f9e0ac1599af18287764e21891ee77f85fc80fc1ed68234ab1a1b18e103e1847R11-R22)

**Parameter and documentation improvements:**

* Extended `TransformationModelLowess::getDefaultParameters` to include new auto-span tuning parameters, with validation and documentation for each.

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

## Summary by CodeRabbit

- **New Features**
  - Auto span selection for Lowess smoothing via cross-validation (LOO/K‑fold), configurable metrics (RMSE, MAE, p90/p95/p99), candidate grids and persistence of chosen span.
  - Cross‑validation utilities: deterministic K‑fold/LOO folding and a 1‑D grid search with tie‑breaking/tolerance policies.
  - Statistical helpers: quantile (sorted/unsorted) and root‑mean‑square‑error utilities.

- **Tests**
  - Unit tests covering auto‑span selection, tie‑breaking/tolerance, cross‑validation, and statistical helpers.

- **Chores**
  - Build updated to include the new cross‑validation module.

### CODE REVIEW

**Comment by singjc** [comment]
@coderabbitai review

File: CHANGELOG [hunk 1]
```diff
@@ -54,7 +54,7 @@ OpenMS Library:
   New Features:
     - added `EnzymaticDigestion.semiSpecificDigestion_()`
     - ProteaseDigestion: add support for semi-specific digestion (#8130)
-
+    - TransformationModelLowess: Add option to automate span selection via cross-validation (##8166)
```
Comment (cbielow) [inline CHANGELOG]:
```suggestion
    - TransformationModelLowess: Add option to automate span selection via cross-validation (#8166)
```

File: src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h [hunk 2]
```diff
@@ -45,6 +45,69 @@ namespace OpenMS
...
+  /**
+    @brief Parse a user-supplied span grid string into candidate values in (0, 1).
+
+    Accepts comma and/or whitespace as separators (e.g. `"0.2, 0.3 0.5"`).
+                           Values `<= 0` or `>= 1` are ignored.
+
+    @param s Input string containing candidate spans.
+    @return Vector of valid candidate spans in (0, 1).
+ */
+  static std::vector<double> parseSpanGridString(const String& s);
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h]: OpenMS Coding Conventions would name this as

Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h]:
```suggestion
  static std::vector<double> parseSpanGridString_(const String& s);
```

File: src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h [hunk 3]
```diff
@@ -45,6 +45,69 @@ namespace OpenMS
...
+                                           double span_max_param,
+                                           int min_neighbors);
+
+  /**
+    @brief Score absolute residuals according to the selected metric.
+
+    Supported @p metric values:
+      - `"rmse"`: root mean square of errors (vs zero)
+      - `"mae"` : mean absolute error
+      - `"p90"`, `"p95"`, `"p99"`: corresponding empirical quantiles of |errors|
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h:96]: please use an `enum class` for this. String arguments are very hard to use from a user perspective.

File: src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h (hunk 2 above)
Comment (cbielow) [reply src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/TransformationModelLowess.h]: also for all other private/protected functions below

File: src/openms/include/OpenMS/MATH/StatisticFunctions.h [hunk 4]
```diff
@@ -196,6 +197,34 @@ namespace OpenMS
       return mean_value / std::distance(begin, end);
     }
 
+    /**
```
Comment (cbielow) [inline src/openms/include/OpenMS/MATH/StatisticFunctions.h]: this is already implemented in StatisticFunctions.h as
`MeanAbsoluteDeviation`.

The name here is a bit misleading as well, since mean absolute error has a different definition to the one used here.
If you want a separate function, then maybe like this:

```
double meanAbsoluteValue(IteratorType begin, IteratorType end)
{
  return MeanAbsoluteDeviation(begin, end, 0.0);
}
```

File: src/openms/include/OpenMS/MATH/StatisticFunctions.h [hunk 5]
```diff
@@ -255,6 +284,72 @@ namespace OpenMS
...
+      @param q      Quantile in [0, 1]
+      @param sorted Whether the input is already sorted; if false, it will be sorted in-place.
+
+      @exception Exception::InvalidRange is thrown if the range is NULL or empty.
+      @exception Exception::InvalidValue is thrown if q is outside [0, 1].
+
+      @ingroup MathFunctionsStatistics
+    */
+    template <typename IteratorType>
+    static double quantile(IteratorType begin, IteratorType end, double q, bool sorted = false)
```
Comment (cbielow) [inline src/openms/include/OpenMS/MATH/StatisticFunctions.h]: bools are not ideal as interface parameters.
Imagine you see
```
  auto val = quantile(vec.begin(), vec.end(), 0.3, false);
```
in code.
What does `false` mean?

Better:
```
  auto val = quantileSortedInput(vec.begin(), vec.end(), 0.3);
 // or
  auto val = quantileUnsortedInput(vec.begin(), vec.end(), 0.3);
```

One function just sorts and calls the other.

File: src/openms/include/OpenMS/ML/CROSSVALIDATION/CrossValidation.h [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    @ingroup MachineLearning
+  */
+  template <typename CandIter, typename TrainEval, typename ScoreFn>
+  static std::pair<typename std::iterator_traits<CandIter>::value_type, double>
+  gridSearch1D(CandIter cbegin, CandIter cend,
+               const std::vector<std::vector<Size>>& folds,
+               TrainEval train_eval,
+               ScoreFn score,
+               double tie_tol = 1e-12,
+               bool prefer_larger = true)
```
Comment (cbielow) [inline src/openms/include/OpenMS/ML/CROSSVALIDATION/CrossValidation.h]: if refactoring is not that easy, consider
```
enum class CANDIDATE
{
  PREFER_LARGER,
  PREFER_ANY
};
```
or similar as a parameter for better readability

File: src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp [hunk 7]
```diff
@@ -8,9 +8,18 @@
...
+#include <OpenMS/ML/CROSSVALIDATION/CrossValidation.h>
+#include <OpenMS/CONCEPT/LogStream.h>
 
 #include <algorithm>
 
+#include <cmath>      
+#include <limits>     
+#include <sstream>    
+#include <iterator>   
+#include <iostream>
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp]: please sort

File: src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp [hunk 8]
```diff
@@ -113,4 +226,81 @@ namespace OpenMS
     params.setValidStrings("extrapolation_type", {"two-point-linear","four-point-linear","global-linear"});
   }
 
-}
+  std::vector<double> TransformationModelLowess::parseSpanGridString(const String& s)
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp]: this already exists:
`ListUtils::create<double>(s)`, although you have to decide on one splitter (space, comma, etc).
The functionality is also a bit unexpected in a class called `TransformationModelLowess`

File: src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp [hunk 9]
```diff
@@ -113,4 +226,81 @@ namespace OpenMS
...
+    std::string cpp = s;
+    for (char& c : cpp) if (c == ',') c = ' ';
+    std::istringstream iss(cpp);
+    double v;
+    while (iss >> v) if (v > 0.0 && v < 1.0) out.push_back(v);
+    return out;
+  }
+
+  std::vector<double> TransformationModelLowess::buildSpanGrid(Size n_pts,
+                                    const String& grid_str,
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp]: pass in `std::vector<double>`, and do the parsing outside. Makes the interface cleaner and more widely usable (e.g. if someone already has a vector<double> ready to go.

File: src/tests/class_tests/openms/source/TransformationModelLowess_test.cpp [hunk 10]
```diff
@@ -141,6 +141,49 @@ START_SECTION((void getParameters(Param& params) const))
...
+    const double x = static_cast<double>(i);
+    data.push_back(std::make_pair(x, 2.0 * x + 1.0));
+  }
+
+  // Params: enable auto-span, provide a small grid, keep everything deterministic
+  Param p;
+  TransformationModelLowess::getDefaultParameters(p);
+  p.setValue("span", 0.0);                       // trigger auto when auto_span=true
+  p.setValue("auto_span", "true");
+  p.setValue("auto_span_grid", "0.3,0.8");       // two candidates
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/TransformationModelLowess_test.cpp:164]: you can directly set `{0.3, 0.8}` (DoubeList) as values. No need to convert from string.

File: src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp [hunk 11]
```diff
@@ -68,6 +77,81 @@ namespace OpenMS
       delta = (xmax_ - xmin_) * 0.01; // automatically determine delta
     }
 
+    // Auto-span options
+    const bool auto_span_flag = params_.getValue("auto_span").toBool();
+    const double span_min_param = (double)params_.getValue("auto_span_min");
+    const double span_max_param = (double)params_.getValue("auto_span_max");
+    const int    min_neighbors  = (int)params_.getValue("auto_min_neighbors");
+    const int    k_folds_param  = (int)params_.getValue("auto_k_folds");
+    const String metric         = params_.getValue("auto_metric").toString();
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp]: to convert from a string metric param to an enum, you can use `Helpers::indexOf` (in EnumHelpers.h), as used here:
https://github.com/OpenMS/OpenMS/blob/develop/src/topp/PercolatorAdapter.cpp#L574

File: src/openms/include/OpenMS/MATH/StatisticFunctions.h (hunk 4 above)
Comment (cbielow) [reply src/openms/include/OpenMS/MATH/StatisticFunctions.h]: please add tests for all new functions.

**Review by cbielow — CHANGES_REQUESTED** [review]
certainly useful changes.
Some comments for streamlining

**Comment by singjc** [comment]
> certainly useful changes. Some comments for streamlining

Thanks for all the comments and suggestions! I addressed mostly all of them I think.

**Comment by timosachsenberg** [comment]
would it make sense to put the 1D grid search along with the grid search that @jpfeuffer wrote for Epifany? 
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/ML/GRIDSEARCH/GridSearch.h

File: src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp [hunk 12]
```diff
@@ -20,75 +29,150 @@
...
+          TransformationModelLowess cv_model(train, p_cv);
+
+          for (Size j : folds_in[f])
+          {
+            const double yhat = cv_model.evaluate(x[j]);
+            const double e    = std::fabs(yhat - y[j]);
+            if (std::isfinite(e)) abs_errs.push_back(e);
+          }
+        }
+      };
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/ANALYSIS/MAPMATCHING/TransformationModelLowess.cpp:142]: is this a valid concern? "gridSearch1D may get only infinite scores or pick a span non-deterministically"

File: src/tests/class_tests/openms/source/CrossValidation_test.cpp [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+  {
+    return errs.empty() ? std::numeric_limits<double>::infinity() : errs.front();
+  };
+
+  TEST_EXCEPTION(Exception::InvalidRange,
+                 CrossValidation::gridSearch1D(empty.begin(), empty.end(),
+                                               folds, train_eval, score));
+}
+END_SECTION
+
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/CrossValidation_test.cpp:252]: are corner cases (e.g. n=1 or n=0) also handled?

File: src/tests/class_tests/openms/source/StatisticFunctions_test.cpp [hunk 14]
```diff
@@ -409,6 +409,67 @@ START_SECTION([EXTRA](template <typename IteratorType> static double quantile(It
 }
 END_SECTION
 
+START_SECTION([EXTRA](template <typename IteratorType> static double quantileSortedInput(IteratorType begin, IteratorType end, double q)))
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/StatisticFunctions_test.cpp]: I am not against having a separate method for sorted and unsorted but would be fine with a single one that assumes data is sorted (e.g. has a PRECONDITION or online check for sortedness)

**Review by timosachsenberg — APPROVED** [review]
nice. only some minor comments

**Comment by jpfeuffer** [comment]
Phew. Yes it would make sense but I think it might be quite some work.
It was kind of hard getting it to work for basic multidimensional cases (checking which simple score of one param set is better than the others).
Adding train/test splits and custom comparators might be a different beast in such a general template recursive (i.e. multidimensional) implementation.

**Comment by timosachsenberg** [comment]
> Phew. Yes it would make sense but I think it might be quite some work. It was kind of hard getting it to work for basic multidimensional cases (checking which simple score of one param set is better than the others). Adding train/test splits and custom comparators might be a different beast in such a general template recursive (i.e. multidimensional) implementation.

Ok then maybe not worth it.

File: src/tests/class_tests/openms/source/CrossValidation_test.cpp (hunk 13 above)
Comment (singjc) [reply src/tests/class_tests/openms/source/CrossValidation_test.cpp:252]: I added a test case for n=1.

File: src/tests/class_tests/openms/source/StatisticFunctions_test.cpp (hunk 14 above)
Comment (singjc) [reply src/tests/class_tests/openms/source/StatisticFunctions_test.cpp]: I removed the unsorted method, and added a precondition to the main method that assumes sortedness.

**Comment by singjc** [comment]
> > Phew. Yes it would make sense but I think it might be quite some work. It was kind of hard getting it to work for basic multidimensional cases (checking which simple score of one param set is better than the others). Adding train/test splits and custom comparators might be a different beast in such a general template recursive (i.e. multidimensional) implementation.
> 
> Ok then maybe not worth it.

I agree that it would make more sense and would be nicer to integrate the gridSearch1D with the current GridSearch module, but it does seem like it would require a bit more thought and work. Maybe for another PR?

**Review by timosachsenberg — APPROVED** [review]
